pgx/src/datum/array.rs

```diff
@@ -69,6 +71,33 @@ impl<'a, T: FromDatum + serde::Serialize> serde::Serialize for Array<'a, T> {
     }
 }
 
+impl<'a, T: FromDatum> Drop for Array<'a, T> {
+    fn drop(&mut self) {
+        if let Array {
+            ptr,
+            raw,
+            datum_palloc: Some(data),
+            elem_slice,
+            ..
+        } = self
+        {
+            // It's just a slice, dropping it doesn't "do" anything, but out of an abundance of caution:
+            mem::drop(elem_slice);
+            // No conflict possible with ditching this.
+            unsafe { pg_sys::pfree(data.as_ptr().cast()) };
+
+            // Now we check for detoasting clones
+            let raw = raw.take().map(|r| r.into_ptr());
+            let ptr = ptr.take();
+            match (ptr, raw) {
+                // SAFETY: if pgx detoasted a clone of this varlena, pfree the clone
+                (Some(p), Some(r)) if r.cast() != p => unsafe { pg_sys::pfree(r.as_ptr().cast()) },
+                _ => (),
```

Comment:
Postgres also **may** allocate when you call `pg_detoast_datum`: the canonical way to check if it did is checking pointer equivalence. 🫠 So this `.take()`s both the varlena pointer from earlier and the RawArray from earlier, and checks that they are both `Some`, mapping RawArray to `*mut ArrayType`, casting from there to `*mut pg_sys::varlena` to compare, then casting again to `*mut c_void` to free it.